### ISSUE TITLE
Implement faster low level integer class

### ISSUE THREAD

**Issue by certik** [sympy/symengine#477]
Currently we use `mpz_class`, but it is possible to speed it up further by automatically switching between a machine integer and a `mpz`. One such implementation is `piranha::integer`. We should either use `piranha::integer` or write our own implementation and use it everywhere instead of `mpz_class`. That will speed things up considerably.

There are couple of decisions to be made:

* `piranha::integer` is using 128 bit machine integer. One can instead use 64bit, which is faster up to 64bit, i.e. it depends on the size of integers that the user needs. For a general CAS use, perhaps 64bit is more common, thus it will be faster for smaller integers, slower than `piranha::integer` up to 128bits, and the same fast for bigger ints (since `mpz` must be used).

* `piranha::integer` is using a union and is strictly following the C++ standard. It is possible to speed things up by not strictly following a C++ standard by using a pointer to `mpz` and overload the upper bit (i.e. requiring the pointer to be always at least 8 bit aligned, which seems to always be the case in practice on all current architectures), and use machine integer instead of the pointer. This is what FLINT does.

Ultimately I think we can implement all of the above, and allow the user to choose, or perhaps pick one default, but have the other implementations in the code, so that the user can use them as building blocks.

**Comment by isuruf** [comment]
I was looking at using `piranha::integer` optionally. We use the `mpz_ptr` from `mpz_class` when using it in `mpz_*` functions, eg:
`mpz_pow_ui(tmp.get_mpz_t(), this->i.get_mpz_t(), other.i.get_ui());`

From what I see, `piranha::integer` gives only a `mpz_srcptr`. Is there a way to do this with piranha?

cc @bluescarni

**Comment by bluescarni** [comment]
@isuruf It is not possible in general to get a non const pointer to an ``mpz_struct``. An ``integer`` instance internally contains either a ``static_integer`` (when the value is small and static storage is active) or an ``mpz_struct`` (when the value is large and dynamic storage is active). You can query which type of storage is currently active with the method ``is_static()``:

http://bluescarni.github.io/piranha/doxygen/classpiranha_1_1mp__integer.html#adf48c526bcf71c389c97bfa23519a33b

You can promote an ``integer`` from static to dynamic storage with the ``promote()`` method:

http://bluescarni.github.io/piranha/doxygen/classpiranha_1_1mp__integer.html#a56bea0563041f2916b2950159baf4d6c

So, after you promote the storage to dynamic, it would be possible to get a non-const pointer to the internal ``mpz_struct``. I can add such a method if you think it would be useful for you (but then you would be forcing the storage to dynamic, maybe a ``demote()`` method would be needed as well to convert back to the static storage representation after you are done with the low level GMP routines).

**Comment by certik** [comment]
The optimal solution is to write a wrapper for each GMP function that we want to use, and provide an implementation that uses machine integers and then use the one depending on if `piranha::integer` is using a machine integer or GMP.

**Comment by bluescarni** [comment]
@certik Keep in mind that Piranha's small integer optimisation stores 2 limbs. So you can do what you propose and it would certainly be the best way to go, but you will need something more sophisticated than just long int arithmetic in order to implement the various GMP functions.

**Comment by certik** [comment]
Essentially the `static_integer` would have to implement all the functionality that's needed.

**Comment by bluescarni** [comment]
@certik Yes that's about right. It's a nice little multiprecision project, it does not require the use of advanced techniques (FFT, Karatsuba, etc.) as the operands will be small but still gives some hands-on experience with MP.

**Comment by isuruf** [comment]
@bluescarni, a method to return a `mpz_ptr` and a `demote()` method would be nice to have, to play with `piranha::integer`s and see if what speedups we can gain.

**Comment by bluescarni** [comment]
@isuruf What about this API:

> ``mpz_struct *integer::mpz_ptr()``: this method will convert ``this`` to dynamic storage (if ``this`` is not already using dynamic storage) and it will return a pointer to the internal ``mpz_struct`` used to represent the integral value.

Or would you like something more fine-grained? The demote woule be something like:

> ``bool integer::demote()``: this method will demote ``this`` to static storage (if ``this`` is not already using static storage). If the value represented by ``this`` is too big to be stored in static storage, ``false`` will be returned and ``this`` will not be modified. Otherwise, ``true`` will be returned.

**Comment by isuruf** [comment]
That's perfect. Thanks.

**Comment by isuruf** [comment]
Something like this should work right?
```
mpz_ptr get_mpz_ptr() {
    if (is_static()) promote();
    return &m_int.g_dy();
}
```
I tried this, but it segfaults.

**Comment by bluescarni** [comment]
@isuruf Mhm I can't see anything wrong. It's a fairly common pattern, here for instance:

https://github.com/bluescarni/piranha/blob/master/src/mp_integer.hpp#L1809

Can you provide a small self-contained example that segfaults?

**Comment by isuruf** [comment]
You are right, that's not the problem. It has something to do with how I use piranha::integers.
Can you take a look at the following program?
https://gist.github.com/isuruf/e93d01dd855e47071063

**Comment by isuruf** [comment]
Results for expand2 benchmarks
With `piranha::integer`, min `809ms`
With `mpz_class`, min `992ms`

So that's a very good improvement. If the above issue in the gist could be fixed, I can send a PR.

**Comment by bluescarni** [comment]
@isuruf I have commented on the gist, reproducing here as well.

The problem is here:

```c++
mpz_srcptr get_mpz_t(const piranha::integer &i) {
    return i.get_mpz_view().get();
}
```

The pointer you get from the ``mpz_view`` class is tied to the lifetime of the originating ``mpz_view`` object. The ``mpz_view`` object is destroyed when the function returns, so you are effectively returning a dangling pointer.

Try this:

```c++
#include <iostream>
#include <gmp.h>
#include <piranha/piranha.hpp>

mpz_ptr get_mpz_t(piranha::integer &i) {
    return i.get_mpz_ptr();
}

//mpz_srcptr get_mpz_t(const piranha::integer &i) {
//    return i.get_mpz_view().get();
//}

unsigned long get_ui(const piranha::integer &i) {
    return mpz_get_ui(i.get_mpz_view());
}

piranha::rational powint(const piranha::integer &th, const piranha::integer &other) {
    if (other < 0)
        return 1/powint(th, -other);
    piranha::integer tmp;
    // CHANGE: here we are using directly mpz_view().
    mpz_pow_ui(get_mpz_t(tmp), th.get_mpz_view(), get_ui(other));
    // Following works
    // mpz_pow_ui(get_mpz_t(tmp), th.get_mpz_view().get(), get_ui(other));
    return piranha::rational(tmp, piranha::integer(1));
}

int main() {
    piranha::integer a(2), b(-1);
    piranha::rational c = powint(a, b);
    return 0;
}
```

The ``mpz_view`` class is implicitly convertible to a const ``mpz_struct`` pointer, so you can use it directly in any GMP function as a ``mpz_srcptr`` argument. The exception is those GMP "functions" which are not actually functions, but macros: macros do not do any type of implicit conversion, so you cannot directly use the ``mpz_view`` class. In that case you need to create the ``mpz_view`` object separately, and then get the pointer to the internal ``mpz_struct`` with the get method. E.g.:

```c++
auto view = n.get_mpz_view();
int sign = mpz_sgn(view.get());
```

``mpz_sign()`` is implemented as a macro, even though it looks like a function. It is documented here:

https://gmplib.org/manual/Integer-Comparisons.html#Integer-Comparisons

The important thing is that the pointer you get from ``view.get()`` is tied to the lifetime of the ``view`` object.

**Comment by isuruf** [comment]
@bluescarni, thanks. That fixes the problems and all tests pass. I mentioned some of the things missing in `piranha::integer` compared to `mpz_class` here https://gitter.im/bluescarni/piranha
